reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java

```diff
@@ -0,0 +1,829 @@
+		@SuppressWarnings("unchecked")
+		ConcurrentLinkedQueue<Slot> q = CONNECTIONS.get(slot.pool);
+		if (q != null) {
+			q.remove(slot);
+		}
+	}
+
+	static final class Borrower extends AtomicBoolean implements Scannable, Subscription, Runnable {
+
+		static final Disposable TIMEOUT_DISPOSED = Disposables.disposed();
+
+		final Duration acquireTimeout;
+		final CoreSubscriber<? super Http2PooledRef> actual;
+		final Http2Pool pool;
+
+		Disposable timeoutTask;
+
+		Borrower(CoreSubscriber<? super Http2PooledRef> actual, Http2Pool pool, Duration acquireTimeout) {
+			this.acquireTimeout = acquireTimeout;
+			this.actual = actual;
+			this.pool = pool;
+			this.timeoutTask = TIMEOUT_DISPOSED;
+		}
+
+		@Override
+		public void cancel() {
+			stopPendingCountdown();
+			if (compareAndSet(false, true)) {
+				pool.cancelAcquire(this);
+			}
+		}
+
+		Context currentContext() {
+			return actual.currentContext();
+		}
+
+		@Override
+		public void request(long n) {
+			if (Operators.validate(n)) {
+				if (!acquireTimeout.isZero()) {
+					timeoutTask = Schedulers.parallel().schedule(this, acquireTimeout.toMillis(), TimeUnit.MILLISECONDS);
+				}
+				pool.doAcquire(this);
+			}
+		}
+
+		@Override
+		public void run() {
+			if (compareAndSet(false, true)) {
+				pool.cancelAcquire(Http2Pool.Borrower.this);
+				actual.onError(new PoolAcquireTimeoutException(acquireTimeout));
+			}
+		}
+
+		@Override
+		@Nullable
+		@SuppressWarnings("rawtypes")
+		public Object scanUnsafe(Attr key) {
+			if (key == Attr.CANCELLED) {
+				return get();
+			}
+			if (key == Attr.REQUESTED_FROM_DOWNSTREAM) {
+				return 1;
+			}
+			if (key == Attr.ACTUAL) {
+				return actual;
+			}
+
+			return null;
+		}
+
+		@Override
+		public String toString() {
+			return get() ? "Borrower(cancelled)" : "Borrower";
+		}
+
+		void deliver(Http2PooledRef poolSlot) {
+			stopPendingCountdown();
+			if (get()) {
+				//CANCELLED or timeout reached
+				poolSlot.invalidate().subscribe(aVoid -> {}, e -> Operators.onErrorDropped(e, Context.empty()));
+			}
+			else {
+				actual.onNext(poolSlot);
+				actual.onComplete();
+			}
+		}
+
+		void fail(Throwable error) {
+			stopPendingCountdown();
+			if (!get()) {
+				actual.onError(error);
+			}
+		}
+
+		void stopPendingCountdown() {
+			timeoutTask.dispose();
+		}
+	}
+
+	static final class BorrowerMono extends Mono<PooledRef<Connection>> {
+
+		final Duration acquireTimeout;
+		final Http2Pool parent;
+
+		BorrowerMono(Http2Pool pool, Duration acquireTimeout) {
+			this.acquireTimeout = acquireTimeout;
+			this.parent = pool;
+		}
+
+		@Override
+		public void subscribe(CoreSubscriber<? super PooledRef<Connection>> actual) {
+			Objects.requireNonNull(actual, "subscribing with null");
+			Borrower borrower = new Borrower(actual, parent, acquireTimeout);
+			actual.onSubscribe(borrower);
+		}
+	}
+
+	static final class Http2PooledRef extends AtomicBoolean implements PooledRef<Connection>, PooledRefMetadata {
```

Comment:
the metrics are not exposed, but we still expose an `acquire` `Mono<PooledRef<Connection>>`. could the various counters, timestamps, etc... methods be improved?

Reply:
`Http2PooledRef` is always invalidated, there is no release functionality, because of that I made most of the metrics NA.

Reply:
ok. I was assuming too much that the `Http2PooledRef` could somehow be presented to other components in the library as a generic `PooledRef` (or even to users). But it is hidden away in the `Http2ConnectionProvider`, so even `lifetime()` information might not be that useful.